File: code/ts/2015/day1/day1.ts
export const getFloor = (instruction: string, currentFloor = 0): number => {
  let floor = currentFloor;
  
  const parens = instruction.split('');
  parens.forEach(paren => {
    if (paren === '(') floor++;
    if (paren === ')') floor--;
  })

  return floor;
}

// Slowest since using substring, (meaning always restarting from index 0 every loop on getFloor)
export const getPositionToFirstEnterBasement = (instruction: string): number => {
  let position = 0;

  for (let i = 1; i <= instruction.length; i++) {
    const floor = getFloor(instruction.substring(0, i));

    if (floor === -1) {
      position = i;
      break;
    }
  }

  return position;
}

// Fastest in benchmark result, but uses additional variable currentFloor
export const getPositionToFirstEnterBasementV2 = (instruction: string): number => {
  let position = 0;
  let currentFloor = 0;

  for (let i = 0; i <= instruction.length; i++) {
    const floor = getFloor(instruction[i], currentFloor)
    
    if (floor === -1) {
      position = i + 1;
      break;
    }

    currentFloor = floor
  }
  
  return position;
}

// Slower than V2, just use a forEach instead of a good 'ol for loop
export const getPositionToFirstEnterBasementV3 = (instruction: string): number => {
  let currentFloor = 0;
  let position = 0;

  const parens = instruction.split('')
  
  parens.forEach((paren, index) => {
    const floor = getFloor(paren, currentFloor)
    
    if (currentFloor !== -1) {
      currentFloor = floor
      position = index + 1
    }
  })

  return position;
}

File: code/py/2015/day1/day1.py
def get_floor (instruction, current_floor = 0):
  floor = current_floor
  
  for paren in instruction:
    if paren == '(': 
      floor += 1
    elif paren == ')':
      floor -= 1

  return floor


def get_position_to_first_enter_basement (instruction):
  position = 0
  current_floor = 0

  length = len(instruction)

  for i in range(length):
    floor = get_floor(instruction[i], current_floor)

    if floor == -1:
      position = i + 1
      break

    current_floor = floor

  return position

def get_position_to_first_enter_basement_v2 (instruction):
  position = 0

  length = len(instruction)

  for i in range(length):
    floor = get_floor(instruction[0:i+1])

    if floor == -1:
      position = i + 1
      break

  return position